src/main/java/org/truffleruby/core/regexp/TruffleRegexpNodes.java

```diff
@@ -400,6 +422,138 @@ protected Object matchInRegion(
+        @Child DispatchNode tRegexGetEndNode;
+
+        @Specialization(guards = "libString.isRubyString(string)")
+        protected Object matchInRegionTRegex(
+                RubyRegexp regexp, Object string, int fromPos, int toPos, boolean atStart, int startPos,
+                @Cached ConditionProfile matchFoundProfile,
+                @Cached ConditionProfile tRegexCouldNotCompileProfile,
+                @Cached ConditionProfile tRegexIncompatibleProfile,
+                @Cached LoopConditionProfile loopProfile,
+                @Cached DispatchNode tRegexExecBytesNode,
+                @Cached DispatchNode tRegexIsMatchNode,
+                @Cached RopeNodes.BytesNode bytesNode,
+                @Cached MatchDataNodes.MatchDataCreateNode matchDataCreateNode,
+                @Cached SelectEncodingNode selectEncodingNode,
+                @Cached TRegexCompileNode tRegexCompileNode,
+                @CachedLibrary(limit = "2") RubyStringLibrary libString) {
+            final Rope rope = libString.getRope(string);
+
+            if (tRegexIncompatibleProfile
+                    .profile(toPos < fromPos || toPos != rope.byteLength() || startPos != 0 || fromPos < 0)) {
+                return fallbackToJoni(regexp, string, fromPos, toPos, atStart, startPos);
+            } else {
+                final RubyEncoding encoding = selectEncodingNode.executeSelectEncoding(regexp, string);
+                final Object compiledRegex = tRegexCompileNode.executeTRegexCompile(regexp, atStart, encoding);
+
+                if (tRegexCouldNotCompileProfile.profile(compiledRegex == nil)) {
+                    return fallbackToJoni(regexp, string, fromPos, toPos, atStart, startPos);
+                }
+
+                final byte[] bytes = bytesNode.execute(rope);
+                final Object regexResult = tRegexExecBytesNode
+                        .call(compiledRegex, "execBytes", getContext().getEnv().asGuestValue(bytes), fromPos);
+
+                final boolean isMatch = (boolean) tRegexIsMatchNode.call(regexResult, "isMatch");
+
+                if (matchFoundProfile.profile(isMatch)) {
+                    final int groupCount = tRegexGroupCount(compiledRegex);
+                    final int[] starts = new int[groupCount];
+                    final int[] ends = new int[groupCount];
+
+                    try {
+                        loopProfile.profileCounted(groupCount);
+
+                        for (int pos = 0; loopProfile.inject(pos < groupCount); pos++) {
+                            starts[pos] = tRegexGetStart(regexResult, pos);
+                            ends[pos] = tRegexGetEnd(regexResult, pos);
+                        }
+                    } finally {
+                        LoopNode.reportLoopCount(this, groupCount);
+                    }
+
+                    return matchDataCreateNode
+                            .executeMatchDataCreate(regexp, dupString(string), starts, ends);
```

Comment:
I think we can just inline that logic here to save one node.

Reply:
Sounds good. You could avoid the intermediary array, too, then and just populate `Region#beg` and `Region#end` directly.